framework/test/gtest/ServiceFactoryTest.cpp

```diff
@@ -201,3 +232,389 @@ TEST_F(ServiceFactoryTest, TestGetServiceReturnsNull)
+    bundleH.GetRegisteredServices();
+  //Test that the # of registered services is seven.
+  ASSERT_EQ(registeredRefs.size(), 7);
+  //Test that First service is bundle scope
+  ASSERT_EQ(registeredRefs[0].GetProperty(Constants::SERVICE_SCOPE).ToString(),
+            Constants::SCOPE_BUNDLE);
+  //Test that Second service is bundle scope
+  ASSERT_EQ(registeredRefs[1].GetProperty(Constants::SERVICE_SCOPE).ToString(),
+            Constants::SCOPE_BUNDLE);
+  //Test that Third service is bundle scope
+  ASSERT_EQ(registeredRefs[2].GetProperty(Constants::SERVICE_SCOPE).ToString(),
+            Constants::SCOPE_BUNDLE);
+  //Test that Fourth service is bundle scope
+  ASSERT_EQ(registeredRefs[3].GetProperty(Constants::SERVICE_SCOPE).ToString(),
+            Constants::SCOPE_BUNDLE);
+  //Test that Fifth service is bundle scope
+  ASSERT_EQ(registeredRefs[4].GetProperty(Constants::SERVICE_SCOPE).ToString(),
+            Constants::SCOPE_BUNDLE);
+  //Test that Sixth service is prototype scope
+  ASSERT_EQ(registeredRefs[5].GetProperty(Constants::SERVICE_SCOPE).ToString(),
+            Constants::SCOPE_BUNDLE);
+  //Test that Seventh service is prototype scope
+  ASSERT_EQ(registeredRefs[6].GetProperty(Constants::SERVICE_SCOPE).ToString(),
+            Constants::SCOPE_PROTOTYPE);
+
+  // Check that a service reference exists
+  const ServiceReferenceU sr1 =
+    context.GetServiceReference("cppmicroservices::TestBundleH");
+  ASSERT_TRUE(sr1);
+  //service is bundle scope
+  ASSERT_EQ(sr1.GetProperty(Constants::SERVICE_SCOPE).ToString(),
+            Constants::SCOPE_BUNDLE);
+
+  InterfaceMapConstPtr service = context.GetService(sr1);
+  ASSERT_TRUE(service);
+  //Must Return at least 1 service object
+  EXPECT_GE(static_cast<int>(service->size()), 1);
+  InterfaceMap::const_iterator serviceIter =
+    service->find("cppmicroservices::TestBundleH");
+  //Find a service object implementing the 'cppmicroservices::TestBundleH' interface
+  ASSERT_NE(serviceIter, service->end());
+  //Test that The service object is valid (i.e. not nullptr)
+  ASSERT_NE(serviceIter->second, nullptr);
+
+  InterfaceMapConstPtr service2 = context.GetService(sr1);
+  //Test that Two service interface maps are equal
+  ASSERT_EQ(*(service.get()), *(service2.get()));
+
+  std::vector<ServiceReferenceU> usedRefs =
+    context.GetBundle().GetServicesInUse();
+  ASSERT_EQ(usedRefs.size(), 1);
+  ASSERT_EQ(usedRefs[0], sr1);
+
+  InterfaceMapConstPtr service3 = bundleH.GetBundleContext().GetService(sr1);
+  ASSERT_NE(service.get(), service3.get());
+
+  // release any service objects before stopping the bundle
+  service.reset();
+  service2.reset();
+  service3.reset();
+
+  bundleH.Stop();
+}
+
+#ifdef US_ENABLE_THREADING_SUPPORT
+
+// test that concurrent calls to ServiceFactory::GetService and ServiceFactory::UngetService
+// don't cause race conditions.
+TEST_F(ServiceFactoryTest, TestConcurrentServiceFactory)
+{
+  auto framework2 = framework;
```

Comment:
is this necessary? why is a copy being made?

Reply:
line 417 : worker_threads.push_back(std::thread([framework2]() requires a variable instead of member ('framework' in this case).
So I can either start a new framework within this test-point or use a copy.